cmd/query/app/otel/otlp_translator.go

```diff
@@ -0,0 +1,353 @@
+	}
+
+	s := &v1.Span{
+		TraceId:           uint64ToTraceID(jSpan.TraceID.High, jSpan.TraceID.Low),
+		SpanId:            uint64ToSpanID(uint64(jSpan.SpanID)),
+		ParentSpanId:      uint64ToSpanID(uint64(jSpan.ParentSpanID())),
+		TraceState:        traceState,
+		Name:              jSpan.GetOperationName(),
+		StartTimeUnixNano: uint64(jSpan.GetStartTime().UnixNano()),
+		EndTimeUnixNano:   uint64(jSpan.GetStartTime().Add(jSpan.GetDuration()).UnixNano()),
+		Events:            jLogsToOTLP(jSpan.GetLogs()),
+		Links:             jReferencesToOTLP(jSpan.GetReferences(), jSpan.ParentSpanID()),
+		Status:            status,
+	}
+	if kind, found := jSpan.GetSpanKind(); found {
+		s.Kind = jSpanKindToInternal(kind)
+		ignoreKeys[tracetranslator.TagSpanKind] = true
+	}
+
+	il := instrumentationLibrary{}
+	if libraryName, ok := tags.FindByKey(conventions.InstrumentationLibraryName); ok {
+		il.name = libraryName.GetVStr()
+		ignoreKeys[conventions.InstrumentationLibraryName] = true
+		if libraryVersion, ok := tags.FindByKey(conventions.InstrumentationLibraryVersion); ok {
+			il.version = libraryVersion.GetVStr()
+			ignoreKeys[conventions.InstrumentationLibraryVersion] = true
+		}
+	}
+	// convert to attributes at the end once not needed attrs are removed
+	attrs := jTagsToOTLP(tags, ignoreKeys)
+	s.Attributes = attrs
+
+	res := resource{}
+	if jSpan.GetProcess() != nil {
+		res.serviceName = jSpan.GetProcess().GetServiceName()
+		res.tags = concatStringTags(jSpan.GetProcess().GetTags())
+		res.resource = jProcessToInternalResource(jSpan.GetProcess())
+	}
+	return s, res, il
+}
+
+// concatStringTags returns key sorted concatenated string tags
+// e.g. keyB=val1,keyA=val2 becomes keyAval2KeyBval1
+func concatStringTags(tags []model.KeyValue) string {
+	keys := make([]string, len(tags))
+	tagMap := make(map[string]string, len(tags))
+	stringTagsLen := 0
+	for i, t := range tags {
+		if t.GetVType() == model.ValueType_STRING {
+			keys[i] = t.GetKey()
+			tagMap[t.GetKey()] = t.GetVStr()
+			stringTagsLen += len(t.GetKey()) + len(t.GetVStr())
+		}
+	}
+	sort.Strings(keys)
+	sBuilder := strings.Builder{}
+	sBuilder.Grow(stringTagsLen)
+	for k, v := range tagMap {
+		sBuilder.WriteString(k)
+		sBuilder.WriteString(v)
+	}
+	return sBuilder.String()
+}
+
+func jProcessToInternalResource(process *model.Process) *resourcev1.Resource {
+	if process == nil {
+		return nil
+	}
+	tags := process.GetTags()
+	if process.GetServiceName() != "" {
+		tags = append(tags, model.String(conventions.AttributeServiceName, process.GetServiceName()))
+	}
+	return &resourcev1.Resource{
+		Attributes: jTagsToOTLP(tags, nil),
+	}
+}
+
+func jTagsToOTLP(tags []model.KeyValue, ignoreKeys map[string]bool) []*commonv1.KeyValue {
+	var kvs []*commonv1.KeyValue
+	for _, tag := range tags {
+		if ignoreKeys[tag.GetKey()] {
+			continue
+		}
+
+		kv := &commonv1.KeyValue{
+			Key:   tag.GetKey(),
+			Value: &commonv1.AnyValue{},
+		}
+		kvs = append(kvs, kv)
+		switch tag.GetVType() {
```

Comment:
How about a default case, to future-proof this?

Reply:
This translates jaeger model to OTLP, I do not expect any modifications, especially to tag types.

Reply:
Hence "future-proof" :-) If the default case is ever reached, you should get a big warning somewhere.